File: iosApp/iosApp/Views/StudyDetails/StudyDetailsView.swift
//
//  StudyDetailsView.swift
//  iosApp
//
//  Created by Daniil Barkov on 22.03.23.
//  Copyright © 2023 Ludwig Boltzmann Institute for
//  Digital Health and Prevention - A research institute
//  of the Ludwig Boltzmann Gesellschaft,
//  Oesterreichische Vereinigung zur Foerderung
//  der wissenschaftlichen Forschung
//  Licensed under the Apache 2.0 license with Commons Clause
//  (see https://www.apache.org/licenses/LICENSE-2.0 and
//  https://commonsclause.com/).
//

import shared
import SwiftUI

struct StudyDetailsView: View {
    @StateObject var viewModel: StudyDetailsViewModel
    private let stringTable = "StudyDetailsView"

    @State var selection: Int = 0
    private let navigationStrings = "Navigation"

    @State private var isObservationListOpen = false

    @EnvironmentObject private var navigationModalState: NavigationModalState

    var body: some View {
        ScrollView {
            VStack(alignment: .leading) {
                Title2(titleText: viewModel.studyDetailsModel?.study.studyTitle ?? "")
                    .padding(.top)
                    .padding(.bottom)

                TaskCompletionBarView(viewModel: TaskCompletionBarViewModel(), progressViewTitle: "tasks_completed")
                    .padding(.bottom, 0.2)

                HStack(alignment: .center) {
                    BasicText(text: "study_duration")

                    Spacer()
                    BasicText(text: (viewModel.studyStart.formattedString()) + " - " + (viewModel.studyEnd.formattedString()),
                              color: Color.more.secondary
                    )
                }.padding(.bottom)

                ExpandableText(viewModel.studyDetailsModel?.study.participantInfo ?? "", "participant_info", lineLimit: 4)
                    .padding(.bottom, 35)

                ExpandableContentWithLink(
                    content: {
                        ScrollView {
                            VStack {
                                ForEach(viewModel.studyDetailsModel?.observations ?? [], id: \.self) { obs in
                                    Button {
                                        navigationModalState.openView(screen: .observationDetails, observationId: obs.observationId)
                                    } label: {
                                        ModuleListItem(observation: obs)
                                            .padding(.bottom)
                                    }
                                }
                            }
                        }
                    },
                    title: { String(localized: "obs_modules") }, expanded: $isObservationListOpen
                ).padding(.top, 0.5)

                Spacer()
            }
        }
        .customNavigationTitle(with: NavigationScreen.studyDetails.localize(), displayMode: .inline)
        .onAppear {
            viewModel.viewDidAppear()
        }
        .onDisappear {
            viewModel.viewDidDisappear()
        }
    }
}

struct StudyDetailsView_Previews: PreviewProvider {
    static var previews: some View {
        StudyDetailsView(viewModel: StudyDetailsViewModel())
    }
}

File: androidApp/src/main/java/io/redlink/more/app/android/activities/studyDetails/StudyDetailsView.kt
/*
 * Copyright LBI-DHP and/or licensed to LBI-DHP under one or more
 * contributor license agreements (LBI-DHP: Ludwig Boltzmann Institute
 * for Digital Health and Prevention -- A research institute of the
 * Ludwig Boltzmann Gesellschaft, Österreichische Vereinigung zur
 * Förderung der wissenschaftlichen Forschung).
 * Licensed under the Apache 2.0 license with Commons Clause
 * (see https://www.apache.org/licenses/LICENSE-2.0 and
 * https://commonsclause.com/).
 */
package io.redlink.more.app.android.activities.studyDetails

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.navigation.NavController
import io.redlink.io.more.app.android.R
import io.redlink.more.app.android.activities.studyDetails.composables.AccordionWithList
import io.redlink.more.app.android.activities.taskCompletion.TaskCompletionBarView
import io.redlink.more.app.android.activities.taskCompletion.TaskCompletionBarViewModel
import io.redlink.more.app.android.extensions.formattedString
import io.redlink.more.app.android.extensions.getStringResource
import io.redlink.more.app.android.extensions.jvmLocalDateTime
import io.redlink.more.app.android.shared_composables.AccordionReadMore
import io.redlink.more.app.android.shared_composables.BasicText
import io.redlink.more.app.android.shared_composables.HeaderTitle
import io.redlink.more.app.android.theme.MoreColors

@Composable
fun StudyDetailsView(
    navController: NavController,
    taskCompletionBarViewModel: TaskCompletionBarViewModel
) {
    val viewModel = remember { StudyDetailsViewModel() }
    val studyInfo by viewModel.coreViewModel.studyModel.collectAsStateWithLifecycle()

    studyInfo?.let {
        Column(
            verticalArrangement = Arrangement.Top,
            modifier = Modifier
                .fillMaxSize()
        ) {
            LazyColumn {
                item {
                    HeaderTitle(title = it.study.studyTitle)
                    Spacer(Modifier.height(12.dp))
                    TaskCompletionBarView(taskCompletionBarViewModel)
                    Spacer(Modifier.height(8.dp))
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.SpaceBetween,
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        if (it.study.start != null && it.study.end != null) {
                            BasicText(text = "${getStringResource(R.string.study_duration)}: ")
                            BasicText(
                                text = "${
                                    it.study.start!!.jvmLocalDateTime().formattedString()
                                } - ${
                                    it.study.end!!.jvmLocalDateTime().formattedString()
                                }",
                                color = MoreColors.Companion.Secondary
                            )
                        }

                    }
                    Spacer(Modifier.height(40.dp))
                    AccordionReadMore(
                        title = getStringResource(R.string.participant_information),
                        description = it.study.participantInfo,
                        modifier = Modifier
                            .fillMaxWidth()
                    )
                    Spacer(Modifier.height(16.dp))

                    AccordionWithList(
                        title = getStringResource(R.string.observation_modules),
                        observations = it.observations,
                        navController = navController
                    )
                }
            }
        }
    }
}